Use map in parser

mohsen1 opened the pull request:
Does this make sense so I spend time clean it up and add test that actually test something?

DeMoorJasper commented:
What exactly changed? Could you add a bt more info on why this change was made?

<sub>Sent with <a href="http://githawk.com">GitHawk</a></sub>

mohsen1 commented on src/Parser.js:
```diff
@@ -5,7 +5,11 @@ const glob = require('glob');
+     * A map of extension names to either path to Asset class or Asset class itself
+     * @type {Map<string, string|Asset}
+     */
+    this.extensions = new Map();
```
@DeMoorJasper this is the core of the change. I'll clean up unnecessary changes momentary

DeMoorJasper replied:
I see you changed it into a map but wondering what issue this fixes?

DeMoorJasper commented:
The only reason I could think of for this PR is performance?

I've made a JSPerf and it's about the same (1% faster)
https://jsperf.com/map-vs-object-11116

DeMoorJasper commented on src/Parser.js:
```diff
@@ -58,18 +62,22 @@ class Parser {
+    // console.log({parser});
     if (typeof parser === 'string') {
-      parser = this.extensions[extension] = require(parser);
+      parser = this.extensions.set(extension, require(parser)).get(extension);
```
If performance is the point of this PR than this actually could be faster if you write it like this:
```Javascript
parser = require(parser);
this.extensions.set(extension, parser);
```

Test case: https://jsperf.com/assign-vs-getting-map

mohsen1 commented:
@DeMoorJasper Thanks for your comments. I cleaned up the code a bit more. 

Not sure how to unit test this. My [first attempt](https://github.com/mohsen1/parcel/blob/21f2bba106fed8454e155d3f844c4dd1d6303591/test/parser.js#L53-L76) didn't really test anything so I removed it.

DeMoorJasper commented:
Not sure why this actually decreased test coverage this much, hope this can be resolved

mohsen1 commented:
Changes from https://github.com/parcel-bundler/parcel/pull/1442 is also here. if coverage is line based we're adding a lot of new lines with those prettier formatted code. Lets merge #1442 first so this change list becomes what it says
